src/front/spv/flow.rs

```diff
@@ -217,33 +217,34 @@ impl FlowGraph {
     }
 
     /// Traverses the flow graph and returns a list of Naga's statements.
-    pub(super) fn to_naga(&self) -> Result<crate::Block, Error> {
+    pub(super) fn convert_to_naga(&mut self) -> Result<crate::Block, Error> {
         self.naga_traverse(node_index(0), None)
     }
 
     fn naga_traverse(
-        &self,
+        &mut self,
         node_index: BlockNodeIndex,
         stop_node_index: Option<BlockNodeIndex>,
     ) -> Result<crate::Block, Error> {
         if stop_node_index == Some(node_index) {
             return Ok(vec![]);
         }
 
-        let node = &self.flow[node_index];
+        let terminator = self.flow[node_index].terminator.clone();
 
-        match node.ty {
-            Some(ControlFlowNodeType::Header) => match node.terminator {
+        match self.flow[node_index].ty {
+            Some(ControlFlowNodeType::Header) => match terminator {
                 Terminator::BranchConditional {
                     condition,
                     true_id,
                     false_id,
                 } => {
                     let true_node_index = self.block_to_node[&true_id];
                     let false_node_index = self.block_to_node[&false_id];
-                    let merge_node_index = self.block_to_node[&node.merge.unwrap().merge_block_id];
+                    let merge_node_index =
+                        self.block_to_node[&self.flow[node_index].merge.unwrap().merge_block_id];
 
-                    let mut result = node.block.clone();
+                    let mut result: crate::Block = self.flow[node_index].block.drain(..).collect();
```

Comment:
Is there a guarantee that `node_index` is only ever going to be processed once?
If it's the case, can we just do something like `let node = self.flow.remove(node_index)` here, and then move the `block` out?
If it's not the case, then I don't understand what's going to happen here: we may try to drain a block that's already empty.

If neither of those is a concern, and the way it's done currently is right, we should do `mem::replace(&mut self.flow[node_index].block, Vec::new())` instead of this `drain` + `collect` pair, since the replace would avoid another heap allocation.

Reply:
As I have written in https://github.com/gfx-rs/naga/issues/691#issuecomment-821100226 I believe we can assume that each node should be processed only once.

I alway forget the trick with mem::replace :) my rust-fu is still weak It seems.

Reply:
Could you clarify why we aren't just removing the node upon traversal?
```
let node = self.flow.remove(node_index);
```

Reply:
I am not sure removing is entirely correct. We visit each node once for cloning statements(so really I use visit word for the purpose of dding statements to IR). But we might still need the nodes for other stuff. For example switch cases ask about accesibility to other cases.